src/probe_modules/module_tcp_synackscan.c

```diff
@@ -38,9 +45,88 @@ static int synackscan_init_perthread(void* buf, macaddr_t *src,
 	return EXIT_SUCCESS;
 }
 
+static int synackscan_make_packet(void *buf, ipaddr_n_t src_ip, ipaddr_n_t dst_ip,
+		uint32_t *validation, int probe_num, __attribute__((unused)) void *arg)
+{
+	struct ether_header *eth_header = (struct ether_header *)buf;
+	struct ip *ip_header = (struct ip*)(&eth_header[1]);
+	struct tcphdr *tcp_header = (struct tcphdr*)(&ip_header[1]);
+	uint32_t tcp_seq = validation[0];
+	uint32_t tcp_ack = validation[2]; //get_src_port() below uses validation 1 internally.
+
+	ip_header->ip_src.s_addr = src_ip;
+	ip_header->ip_dst.s_addr = dst_ip;
+
+	tcp_header->th_sport = htons(get_src_port(num_ports,
+				probe_num, validation));
+	tcp_header->th_seq = tcp_seq;
+	tcp_header->th_ack = tcp_ack;
+	tcp_header->th_sum = 0;
+	tcp_header->th_sum = tcp_checksum(sizeof(struct tcphdr),
+			ip_header->ip_src.s_addr, ip_header->ip_dst.s_addr, tcp_header);
+
+	ip_header->ip_sum = 0;
+	ip_header->ip_sum = zmap_ip_checksum((unsigned short *) ip_header);
+
+	return EXIT_SUCCESS;
+}
+
+static int synackscan_validate_packet(const struct ip *ip_hdr, uint32_t len,
+		__attribute__((unused))uint32_t *src_ip,
+		uint32_t *validation)
+{
+	if (ip_hdr->ip_p != IPPROTO_TCP) {
+		return 0;
+	}
+	if ((4*ip_hdr->ip_hl + sizeof(struct tcphdr)) > len) {
+		// buffer not large enough to contain expected tcp header
+		return 0;
+	}
+	struct tcphdr *tcp = (struct tcphdr*)((char *) ip_hdr + 4*ip_hdr->ip_hl);
+	uint16_t sport = tcp->th_sport;
+	uint16_t dport = tcp->th_dport;
+	// validate source port
+	if (ntohs(sport) != zconf.target_port) {
+		return 0;
+	}
+	// validate destination port
+	if (!check_dst_port(ntohs(dport), num_ports, validation)) {
+		return 0;
+	}
+
+	// Note: The traditional flow of returning 0 on wrong and then returning 1
+	// at the end made this code an unreadable mess. So we're going a bit
+	// against the grain
+
+	if (htonl(tcp->th_ack) == htonl(validation[0]) + 1) {
+		// ACK = SEQ + 1 is OK
+		return 1;
+	} else if (tcp->th_flags & TH_RST) {
+		// If we have a reset flag, we can have a few cases.
+		if (htonl(tcp->th_seq) == htonl(validation[2]) + 0) {
+			// If SEQ = ACK RST is valid
+			return 1;
+		} else if (htonl(tcp->th_seq) == htonl(validation[2]) + 1) {
+			// If SEQ = ACK + 1 RST is also valid
+			return 1;
+		} else {
+			// All other cases the RST is invalid.
+			return 0;
+		}
+	} else {
+		// All other options are invalid.
+		return 0;
+	}
+
+	// DO NOT ADD ANY CHECKS HERE.
+	assert(NULL);
```

Comment:
In general, I'm not sure what the "success" conditions are supposed to be here. What are they?